lib/PreviewProvider.js

```diff
@@ -0,0 +1,137 @@
+/* globals Task */
+"use strict";
+
+const {Cu} = require("chrome");
+const ss = require("sdk/simple-storage");
+const {setTimeout} = require("sdk/timers");
+const {siteFilter} = require("content-src/lib/filters");
+
+const EMBEDLY_PROXY_MAX_LINKS = 25; // number of links embedly proxy accepts per request
+const EMBEDLY_ENDPOINT = "https://embedly-proxy.dev.mozaws.net/v2/extract";
+const CLEAR_CACHE_TIMEOUT = 86400000; // 24 hours to clear cache
+
+Cu.importGlobalProperties(["fetch"]);
+Cu.import("resource://gre/modules/Task.jsm");
+
+function PreviewProvider() {
+  this.init();
+}
+
+PreviewProvider.prototype = {
+  // TODO: santize, filter, and dedupe urls
+  /**
+    * Clean up cache
+    */
+  cleanUpCache: Task.async(function*() {
+    setTimeout(() => {
+      let currentTime = Date.now();
+      Object.keys(ss.storage.embedlyData).forEach(item => {
+        if (currentTime - ss.storage.embedlyData[item].accessTime > CLEAR_CACHE_TIMEOUT) {
+          delete ss.storage.embedlyData[item];
+        }
+      });
+      this.cleanUpCache();
+    }, CLEAR_CACHE_TIMEOUT);
+  }),
+
+  /**
+    * Dedupe urls
+    */
+  dedupeURLs(items) {
+    let createKey =  (item => item);
+    let compare = (() => false);
+    const valueMap = new Map();
+    items.forEach(value => {
+      const key = createKey(value);
+      if (!valueMap.has(key) || compare(valueMap.get(key), value)) {
+        valueMap.set(key, value);
+      }
+    });
+    return Array.from(valueMap.values());
+  },
+
+  /**
+    * Filter cached links out, and store their embedly data
+    */
+  getCachedLinks(links) {
+    let cachedLinks = [];
+    links.forEach(link => {
+      if (link && ss.storage.embedlyData[link.url] && link.url === ss.storage.embedlyData[link.url].url) {
+        cachedLinks.push(ss.storage.embedlyData[link.url]);
+      }
+    });
+    return cachedLinks;
+  },
+
+  /**
+    * Prime the cache with embedly data
+    */
+  primeCache: Task.async(function*(links) {
+    this.saveNewLinks(links);
+  }),
+
+  /**
+   * Filter out new links and request them from embedly
+   */
+  saveNewLinks: Task.async(function*(links) {
+    let newLinks = links.filter(link => link && !ss.storage.embedlyData[link.url]);
+    let requestQueue = [];
+    while (newLinks.length !== 0) {
+      // we have some new links we need to fetch the embedly data for, put them on the queue
+      requestQueue.push(newLinks.splice(0, EMBEDLY_PROXY_MAX_LINKS));
+    }
+    // for each bundle of 25 links, spawn off a new request to embedly
+    requestQueue.forEach(requestBundle => {
+      Task.spawn(function*() {
```

Comment:
There's a design we could do to make this better, but the current implementation can suffice: it is much better than the current situation.

With the code in this PR, `saveNewLinks` will potentially result in an async `fetchAndCache` task, depending on what the filter returns.

The idea behind a new design would be to have a pool of fetchers, so we can cap the number of outbound requests by a set number.

A way to implement this would be to have a queue of links that each call to `saveNewLinks` would insert into. Upon inserting, `saveNewLinks` would be careful not to duplicate canonicalized URLs in the queue.

Asynchronous "threads of executions" check this queue periodically and perform `fetchAndCache` queries. If the queue is empty, the threads of execution "sleep". Of course, they don't really sleep, they timeout for a short amount of time.

Basically, the idea is a common pattern called a "job queue".